### ISSUE TITLE
BUG: transaction lost calling a distributed procedure in Citus 12.1.3 (but 12.1.0 works)

### ISSUE THREAD

**Issue by DS-AdamMilazzo** [citusdata/citus#7597]
In Citus 12.1.3, when calling a distributed procedure in a batch of SQL statements, the CALL statement appears to execute in a different transaction from the rest of the statements in the batch. This behavior is not observed in Citus 12.1.0.

### Repro steps ###
First, perform this setup:
```SQL
CREATE TABLE t (p INT NOT NULL, i INT NOT NULL, PRIMARY KEY (p, i));
SELECT create_distributed_table('t', 'p');
CREATE PROCEDURE f(_p INT, _i INT) LANGUAGE plpgsql AS
    $$ BEGIN SELECT i / 0 FROM t WHERE p = _p AND i = _i; END; $$;
SELECT create_distributed_function('f(INT,INT)', distribution_arg_name := '_p', colocate_with := 't');
```

Second, call the procedure in a parameterized batch of SQL statements:
```SQL
INSERT INTO t (p, i) VALUES (@0, @1);
CALL f(1, @1);
```
In this batch, I used parameters `@0` = 1 and `@1` = 2. The batch was sent with the Npgsql library, but any library that allows sending parameterized batches should be able to reproduce this bug. (Don't try putting the statements into a SQL procedure and calling it. That's not likely to reproduce the problem. Use a client library to send the parameterized batch.)

**Expected behavior:**
1. A tuple should be inserted into table t.
2. Procedure `f(1, 2)` should be called, and fail with a division by zero error.
3. The transaction should roll back.

**Actual behavior:**
1. A tuple is inserted into table t.
2. Procedure `f` is _not_ called. Instead, an error occurs: "42P02: there is no parameter $1"
3. The INSERT statement is still committed (i.e. the inserted tuple is not removed).

### Additional Comments ###
- The problem happens in Citus 12.1.3 but not 12.1.0, I _believe_.
- The problem only happens if it's a distributed function. If you don't call `create_distributed_function`, the problem does not occur.
- It doesn't matter whether the function `f` is supposed to succeed or fail. It never gets called in the first place.
- I previously reported #7242 which was also related to distributed functions. It may or may not be related.
- This is a fairly serious problem that I hope gets fixed soon...

**Comment by emelsimsek** [comment]
I have  tried to reproduce this using the below python code. However on both 12.1.0 and 12.1.3, procedure f is called, it throws division by 0 exception and nothing is inserted to the table t.  Is anything missing in this code?

```
    cur = conn.cursor()

    sql = "CREATE TABLE t (p INT NOT NULL, i INT NOT NULL, PRIMARY KEY (p, i));"
    cur.execute(sql);

    sql = "SELECT create_distributed_table('t', 'p');"
    cur.execute(sql);

    conn.commit();

    sql = """ CREATE PROCEDURE f(_p INT, _i INT) LANGUAGE plpgsql AS
    $$ BEGIN SELECT i / 0 FROM t WHERE p = _p AND i = _i; END; $$;"""

    cur.execute(sql);

    conn.commit();

    sql = "SELECT create_distributed_function('f(INT,INT)', distribution_arg_name := '_p', colocate_with := 't');"
    cur.execute(sql);

    conn.commit();

    sql = "INSERT INTO t (p, i) VALUES (%s, %s); CALL f(1, %s);"

    cur.execute(sql,(1,2,2))
    conn.commit();
```

**Comment by DS-AdamMilazzo** [comment]
I am not familiar with the python client, but I think there are two problems with that code.
1. I think that by default, the python client wraps your commands in transactions. You should switch it to autocommit mode.
2. Ideally you would use two parameters, as in `INSERT INTO t (p, i) VALUES (@0, @1); CALL f(1, @1);` rather than three. Using three will likely result in a different error.

**Comment by DS-AdamMilazzo** [comment]
Any luck reproducing this?

**Comment by m3hm3t** [comment]
I was able to replicate the issue with Npgsql on both Citus versions 12.1.3 and 12.1.0. The behavior described above occurred as expected.

Here is the code I executed:

```
using System;
using Npgsql;

class Program
{
    static void Main()
    {
        // Update the connection string with the port number
        var connString = "Host=localhost;Port=9700;Database=citus";

        // Establish the connection
        using (var connection = new NpgsqlConnection(connString))
        {
            connection.Open();

            // Create a command
            using (var cmd = connection.CreateCommand())
            {
                // Command text
                cmd.CommandText = "INSERT INTO t (p, i) VALUES (@0, @1); CALL f(1, @1);";
                
                // Add parameters
                cmd.Parameters.Add(new NpgsqlParameter("@0", 1));
                cmd.Parameters.Add(new NpgsqlParameter("@1", 2));
                
                // Execute the command
                cmd.ExecuteNonQuery();
            }
        }
    }
}
```


Additionally, I attempted to simulate the same behavior using another library:

```
import psycopg2

# Establish the connection
connection = psycopg2.connect(
    dbname="citus", 
    host="localhost", 
    port="9700"
)

# Enable autocommit mode
connection.autocommit = True

# Execute the combined command with implicit transaction handling
try:
    with connection.cursor() as cursor:
        # Combined command: INSERT INTO t followed by CALL f
        cursor.execute(
            """
            INSERT INTO t (p, i) VALUES (%s, %s);
            CALL f(1, %s);
            """, 
            (1, 2, 2)
        )

except Exception as e:
    print(f"Error during execution: {e}")
finally:
    connection.close()
```

```
citus@0e40d9ee7522:/data$ python3 script.py
Error during execution: division by zero
CONTEXT:  while executing command on localhost:9702
SQL statement "SELECT i / 0 FROM t WHERE p = _p AND i = _i"
PL/pgSQL function f(integer,integer) line 1 at SQL statement
```
In this case, the behavior was as expected.

My experience with Npgsql is somewhat limited, but here’s my take on the issue. It seems like the INSERT works and gets committed right away because of autocommit mode. However, the CALL fails on the client side due to a parameter binding problem in Npgsql. This error happens before the distributed transaction can complete on the worker nodes, so Citus doesn’t roll back the INSERT.

**Comment by DS-AdamMilazzo** [comment]
I don't know what
```
        cursor.execute(
            """
            INSERT INTO t (p, i) VALUES (%s, %s);
            CALL f(1, %s);
            """, 
            (1, 2, 2)
```
does exactly, e.g. whether it's going to use actual SQL-level parameters for the values 1, 2, 2. The [documentation](https://www.psycopg.org/docs/usage.html#passing-parameters-to-sql-queries) is not clear on this point, either. It looks like it may well be passing literals.

Anyway, if it was due to a client-side problem in Npgsql, that would raise the question of why the problem doesn't occur if you don't call `create_distributed_function`. I.e. it only occurs with Citus distributed functions.

**Comment by DS-AdamMilazzo** [comment]
I don't agree that it's a client-side problem, but it only happens when specifying parameters at the SQL level. Clients that transform parameters into literals will not reproduce the problem. Running the function without parameters will not reproduce the problem. Parameter binding works in Npgsql for this function but only if you didn't previously call `create_distributed_function` on it. Npgsql will not know whether the function has been distributed. The same request, which works, only starts to fail after calling `create_distributed_function`.

According to my original comments:

- You don't even need the INSERT statement to trigger the error. Any CALL to a distributed procedure with batch parameters (outside of an explicit transaction) will fail! E.g. if the batch is simply `CALL f(1, @0);` it will fail.
- The problem only happens if it's a distributed function. If you don't call `create_distributed_function`, the problem does not occur.
- The problem does not occur when an explicit transaction is used, e.g. `BEGIN; INSERT ...; CALL ...; COMMIT;`
- The problem does not occur if the CALL does not have batch parameters, e.g. `CALL f(1, 2);`
- It doesn't matter whether the function `f` is supposed to succeed or fail. It never gets called in the first place.

There was a similar bug reported previously - #7242 - which was also related to parameterized distributed functions called outside explicit transactions. It may be related, and it was a server-side problem.

**Comment by m3hm3t** [comment]
@DS-AdamMilazzo Do you have existing logs with log_min_messages and citus.work_min_messages to DEBUG1 (and maybe DEBUG4)? If not, I will try to generate them.

**Comment by ihalatci** [comment]
**12.1.14 release triage — deferred (label intentionally retained).**

Reviewed while assembling the 12.1.14 patch. The only PR ever linked to this issue is
#8064, which was a **draft and was closed without being merged**. There is no fix on
`main`, so there is nothing to cherry-pick onto `release-12.1`.

Keeping the `cherry-pick-12.1` label so this stays in the backport queue. It becomes
actionable once a fix lands on `main`.

_(No action needed from the reporter — this is a release-process note.)_
